lib/avalon/audit_objects.rb

```diff
@@ -0,0 +1,50 @@
+module Avalon
+  module AuditObjects
+
+    def self.run
+      bad_objects = []
+      ActiveFedora::Base.all.each do |object|
+        bad_objects << object unless object.valid?
+      end
+      return unless bad_objects.any?
+      AuditObjectsMailer.error_report(bad_objects, error_stats(bad_objects)).deliver_now
+    end
+
+    def self.error_stats(bad_objects)
+      # Each entry has keys :heading and :value to render as a table
+
+      # Total errors
+      stats = [ { heading: I18n.t('audit.total_header_html').html_safe,
+                  value: bad_objects.count } ]
+      object_class_counts = {}
+      collection_counts = {}
+
+      bad_objects.each do |object|
+        # Count occurences for each object class
+        object_class_counts[object.class] ||= 0
+        object_class_counts[object.class] += 1
+
+        # Count occurences for each object class
+        if object.respond_to?(:collection)
+          collection_counts[object.collection.name] ||= 0
+          collection_counts[object.collection.name] += 1
+        end
+      end
+      object_class_counts.each do |klass, count|
+        stats << {
+          heading: I18n.t('audit.class_total_header_html', class: klass).html_safe,
+          value: count
+        }
+      end
+      collection_counts.each do |collection, count|
+        stats << {
+          heading: I18n.t('audit.collection_total_header_html',
+                          collection: collection).html_safe,
+          value: count
+        }
+      end
```

Comment:
For N objects failing validation, we end up performing 3N iterations throughout this method. We could fold this all together into one block by skipping the creation of this intermediate count hashes and updating the stats array directly in the `bad_object.each` block, which is probably a net win in runtime?

Reply:
Think I'm going to leave this one. Will add a TODO to explore optimization -- the 3 or 4 N iterations will be dwarfed by the preceding fedora fetches, and this code is run via cron (not user-facing).

Reply:
Fair enough.